### PR TITLE
numbers.parser refactoring

### PR DESCRIPTION
**Pull request by stepancheg**
- Replace copy-paste with generic code
- Use correct C types to clay types mapping
- Rewrite tests to use test framework
- Add more tests

### CODE REVIEW

File: lib-clay/numbers/parser/parser.clay [hunk 1]
```diff
@@ -9,8 +9,51 @@ import sequences.*;
...
+    parseInt(String(s), intType);
+
+[S when String?(S)]
+overload parseFloat(s:S, floatType) =
+    parseFloat(String(s), floatType);
+
+[S when String?(S)]
+overload parseImag(s:S, imagType) =
+    parseImag(String(s), imagType);
+
```
Comment (jckarter) [inline lib-clay/numbers/parser/parser.clay]: What do you think about folding these into a single `parse(s, T)` generic?

Comment (stepancheg) [inline lib-clay/numbers/parser/parser.clay]: I agree, but I prefer `parse(T, s)`. It is closer to Haskell parameter ordering style (where partial application is possible: `parseInt = parse Int`).

Comment (jckarter) [inline lib-clay/numbers/parser/parser.clay]: I'd normally agree, but `parse` could also be generalized to be variadic and parse multiple objects like `scanf` or `istream::operator>>`, as in `parse(s, Int, String, Int) : Int, String, Int`, and it seems more natural to put the output types after the input stream in that case.

Comment (stepancheg) [inline lib-clay/numbers/parser/parser.clay]: I think that parse function should not be an overloading, but rather something like:

```
parseSpaceSeparated(s, Int, String, Int) : Int, String, Int
```

Comment (jckarter) [inline lib-clay/numbers/parser/parser.clay]: Sure, that works too. In that case, `parse(T, s)` is better for the single-object case.

Comment (stepancheg) [inline lib-clay/numbers/parser/parser.clay]: Should I also kill overloads like

```
[S when String?(S)]
forceinline overload UInt8(s:S) = ...
```

?

Comment (jckarter) [inline lib-clay/numbers/parser/parser.clay]: That'd be great. It'd take care of #116.

**Comment by stepancheg** [comment]
Removed `parseXxx` functions and `IntXxx` constructors.

**Comment by jckarter** [comment]
Looks good, thanks. Are there any other `T(String)` constructors in the library that could be changed to `parse` overloads?

**Comment by stepancheg** [comment]
@jckarter I don't know. But if there are, `parse` definition should be moved to somewhere like `core.parser`?

**Comment by jckarter** [comment]
Indeed, moving it to a new module is a good idea.

**Pull request merged**